README.md
# ssh-hardening (Chef cookbook)

[![Supermarket](http://img.shields.io/cookbook/v/ssh-hardening.svg)][1]
[![Build Status](http://img.shields.io/travis/TelekomLabs/chef-ssh-hardening.svg)][2]
[![Code Coverage](http://img.shields.io/coveralls/TelekomLabs/chef-ssh-hardening.svg)][3]
[![Dependencies](http://img.shields.io/gemnasium/TelekomLabs/chef-ssh-hardening.svg)][4]

## Description

This cookbook provides secure ssh-client and ssh-server configurations.

## Requirements

* Opscode chef

## Attributes

* `['network']['ipv6']['enable']` - true if IPv6 is needed
* `['ssh']['cbc_required']` - true if CBC for ciphers is required. This is usually only necessary, if older M2M mechanism need to communicate with SSH, that don't have any of the configured secure ciphers enabled. CBC is a weak alternative. Anything weaker should be avoided and is thus not available.
* `['ssh']['weak_hmac']` - true if weaker HMAC mechanisms are required. This is usually only necessary, if older M2M mechanism need to communicate with SSH, that don't have any of the configured secure HMACs enabled. 
* `['ssh']['weak_kex']` - true if weaker Key-Exchange (KEX) mechanisms are required. This is usually only necessary, if older M2M mechanism need to communicate with SSH, that don't have any of the configured secure KEXs enabled.
* `['ssh']['allow_root_with_key']` - `false` to disable root login altogether. Set to `true` to allow root to login via key-based mechanism.
* `['ssh']['ports']` - ports to which ssh-server should listen to and ssh-client should connect to
* `['ssh']['listen_to']` - one or more ip addresses, to which ssh-server should listen to. Default is empty, but should be configured for security reasons!
* `['ssh']['remote_hosts']` - one or more hosts, to which ssh-client can connect to. Default is empty, but should be configured for security reasons!
* `['ssh']['allow_tcp_forwarding']` - `false` to disable TCP Forwarding. Set to `true` to allow TCP Forwarding
* `['ssh']['allow_agent_forwarding']` - `false` to disable Agent Forwarding. Set to `true` to allow Agent Forwarding
* `['ssh']['use_pam']` - `false` to disable pam authentication

## Data Bags

**DEPRECATION WARNING**: Support for managing authorized_keys for the root account will be removed from this cookbook in the next major release. Please use alternative cookbooks for that.

This cookbook used to handle authorized keys for the root user, but that support will be removed in the next major release. Use other cookbooks to set up your users.

### Old behaviour:

Have users in your `data_bag/users/` directory. This cookbook looks for users inside this folder with a `ssh_rootkey`.

Example: 

First you have to find out the ssh-key of the user you want to allow. A typical example for this is

    cat ~/.ssh/id_rsa.pub

If that folder doesn't exist or you don't know what this is all about, please read a SSH tutorial for your blend of operating system first.

You can now add this key to the data bag. Example for dada:

Example for `data_bags/users/dada.json`

    {
      "id" : "dada",
      // ... other stuff ...
      "ssh_rootkey" : "ssh-rsa AAAA....mail.com"
    }

You can then access

    ssh dada@yourmachines


## Usage

Add the recipes to the run_list:
    
    "recipe[ssh]"

This will install ssh-server and ssh-client. You can alternatively choose only one via:

    "recipe[ssh::server]"
    "recipe[ssh::client]"

Configure attributes:

    "ssh" : {
      "listen_to" : "10.2.3.4"
    }

**The default value for `listen_to` is `0.0.0.0`. It is highly recommended to change the value.**

## Local Testing

For local testing you can use vagrant and Virtualbox of VMWare to run tests locally. You will have to install Virtualbox and Vagrant on your system. See [Vagrant Downloads](http://downloads.vagrantup.com/) for a vagrant package suitable for your system. For all our tests we use `test-kitchen`. If you are not familiar with `test-kitchen` please have a look at [their guide](http://kitchen.ci/docs/getting-started).

Next install test-kitchen:

```bash
# Install dependencies
gem install bundler
bundle install

# Do lint checks
bundle exec rake lint

# Fetch tests
bundle exec thor kitchen:fetch-remote-tests

# fast test on one machine
bundle exec kitchen test default-ubuntu-1204

# test on all machines
bundle exec kitchen test

# for development
bundle exec kitchen create default-ubuntu-1204
bundle exec kitchen converge default-ubuntu-1204
```

For more information see [test-kitchen](http://kitchen.ci/docs/getting-started)

## FAQ / Pitfalls

**I can't log into my account. I have registered the client key, but it still doesn't let me it.**

If you have exhausted all typical issues (firewall, network, key missing, wrong key, account disabled etc.), it may be that your account is locked. The quickest way to find out is to look at the password hash for your user:

    sudo grep myuser /etc/shadow

If the hash includes an `!`, your account is locked:

    myuser:!:16280:7:60:7:::

The proper way to solve this is to unlock the account (`passwd -u myuser`). If the user doesn't have a password, you should can unlock it via:

    usermod -p "*" myuser

Alternatively, if you intend to use PAM, you enabled it via `['ssh']['use_pam'] = true`. PAM will allow locked users to get in with keys.


**Why doesn't my application connect via SSH anymore?**

Always look into log files first and if possible look at the negotation between client and server that is completed when connecting.

We have seen some issues in applications (based on python and ruby) that are due to their use of an outdated crypto set. This collides with this hardening module, which reduced the list of ciphers, message authentication codes (MACs) and key exchange (KEX) algorithms to a more secure selection.

If you find this isn't enough, feel free to activate `['ssh']['cbc_required']` for ciphers, `['ssh']['weak_hmac']` for MACs, and `['ssh']['weak_kex']` for KEX.

## Contributors + Kudos

* Dominik Richter
* Christoph Hartmann
* Patrick Meier
* Bernhard K. Weisshuhn

This cookbook is mostly based on guides by:

* [NSA: Guide to the Secure Configuration of Red Hat Enterprise Linux 5](http://www.nsa.gov/ia/_files/os/redhat/rhel5-pamphlet-i731.pdf)
* [Deutsche Telekom, Group IT Security, Security Requirements (German)](http://www.telekom.com/static/-/155996/7/technische-sicherheitsanforderungen-si)

Thanks to all of you!!

## Contributing

See [contributor guideline](CONTRIBUTING.md).

## License and Author

* Author:: Dominik Richter <dominik.richter@googlemail.com>
* Author:: Deutsche Telekom AG

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


[1]: https://supermarket.getchef.com/cookbooks/ssh-hardening
[2]: http://travis-ci.org/TelekomLabs/chef-ssh-hardening
[3]: https://coveralls.io/r/TelekomLabs/chef-ssh-hardening
[4]: https://gemnasium.com/TelekomLabs/chef-ssh-hardening


CHANGELOG.md
# Changelog

## unreleased

* deprecated: Manging authorized_keys for root via attributes `ssh_rootkey` and  `ssh_rootkeys` in the `users` data bag has been deprecated and emits a waning when used. Support will be removed in 2.x.

## 1.0.1

* feature: cipher, macs and key exchange algorithms are now correctly detected on
  ubuntu 12.04+14.04, centos/oracle/redhat 6.4+6.5, debian 6+7
* feature: UsePAM can now be configured. Locked accounts may not get access via SSH
  if UsePAM is disabled (which is the default)
* feature: AllowTcpForwarding is now configurable. It is safe to set it if the user
  has a login shell anyway
* improvement: introduced rubocop+foodcritic for linting. As a result, there has been
  a long list of cleanups and fixes to make this project looking well-rounded again
* bugfixes: incorrect crypto-configuration on red-hat based systems and debian

## 1.0.0

* imported ssh hardening project and updated to current version with full test suite


spec/recipes/client_spec.rb
# encoding: UTF-8
#
# Copyright 2014, Deutsche Telekom AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'spec_helper'

describe 'ssh-hardening::client' do

  # converge
  cached(:chef_run) do
    ChefSpec::ServerRunner.new.converge(described_recipe)
  end

  it 'installs openssh-client' do
    expect(chef_run).to install_package('openssh-client')
  end

  it 'creates the directory /etc/ssh' do
    expect(chef_run).to create_directory('/etc/ssh')
      .with(mode: '0755')
      .with(owner: 'root')
      .with(group: 'root')
  end

  it 'creates /etc/ssh/ssh_config' do
    expect(chef_run).to create_template('/etc/ssh/ssh_config')
      .with(owner: 'root')
      .with(group: 'root')
      .with(mode: '0644')
  end

  it 'disables weak hmacs' do
    expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
      .with_content(/MACs [^#]*\bhmac-sha1\b/)
  end

  it 'disables weak kexs' do
    expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
      .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
  end

  it 'disables cbc ciphers' do
    expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
      .with_content(/Ciphers [^#]*-cbc\b/)
  end

  it 'enables ctr ciphers' do
    expect(chef_run).to render_file('/etc/ssh/ssh_config')
      .with_content(/Ciphers [^#]*\baes128-ctr\b/)
      .with_content(/Ciphers [^#]*\baes192-ctr\b/)
      .with_content(/Ciphers [^#]*\baes256-ctr\b/)
  end

  context 'with weak hmacs enabled' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['ssh']['weak_hmac'] = true
      end.converge(described_recipe)
    end

    it 'allows weak hmacs' do
      expect(chef_run).to render_file('/etc/ssh/ssh_config')
        .with_content(/MACs [^#]*\bhmac-sha1\b/)
    end

    it 'still does not allow weak kexs' do
      expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
        .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
    end

    it 'still doss not allow cbc ciphers' do
      expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
        .with_content(/Ciphers [^#]*-cbc\b/)
    end
  end

  context 'with weak kexs enabled' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['ssh']['weak_kex'] = true
      end.converge(described_recipe)
    end

    it 'allows weak kexs' do
      expect(chef_run).to render_file('/etc/ssh/ssh_config')
        .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
    end

    it 'still does not allow weak macs' do
      expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
        .with_content(/MACs [^#]*\bhmac-sha1\b/)
    end

    it 'still does not allow cbc ciphers' do
      expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
        .with_content(/Ciphers [^#]*-cbc\b/)
    end
  end

  context 'with cbc required' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['ssh']['cbc_required'] = true
      end.converge(described_recipe)
    end

    it 'allows cbc ciphers' do
      expect(chef_run).to render_file('/etc/ssh/ssh_config')
        .with_content(/Ciphers [^#]*\baes256-cbc\b/)
        .with_content(/Ciphers [^#]*\baes192-cbc\b/)
        .with_content(/Ciphers [^#]*\baes128-cbc\b/)
    end

    it 'still does not allow weak macs' do
      expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
        .with_content(/MACs [^#]*\bhmac-sha1\b/)
    end

    it 'still does not allow weak kexs' do
      expect(chef_run).not_to render_file('/etc/ssh/ssh_config')
        .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
    end

    it 'still enables ctr ciphers' do
      expect(chef_run).to render_file('/etc/ssh/ssh_config')
        .with_content(/Ciphers [^#]*\baes128-ctr\b/)
        .with_content(/Ciphers [^#]*\baes192-ctr\b/)
        .with_content(/Ciphers [^#]*\baes256-ctr\b/)
    end
  end

  context 'chef-solo' do
    cached(:chef_run) do
      ChefSpec::SoloRunner.new.converge(described_recipe)
    end

    it 'does not raise an error' do
      expect { chef_run }.not_to raise_error
    end
  end
end


spec/recipes/server_spec.rb
# encoding: UTF-8
#
# Copyright 2014, Deutsche Telekom AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'spec_helper'

describe 'ssh-hardening::server' do

  # converge
  cached(:chef_run) do
    ChefSpec::ServerRunner.new.converge(described_recipe)
  end

  it 'installs openssh-server' do
    expect(chef_run).to install_package('openssh-server')
  end

  it 'creates /etc/ssh/sshd_config' do
    expect(chef_run).to create_template('/etc/ssh/sshd_config')
      .with(mode: '0600')
      .with(owner: 'root')
      .with(group: 'root')
  end

  it 'enables the ssh server' do
    expect(chef_run).to enable_service('sshd')
  end

  it 'starts the server' do
    expect(chef_run).to start_service('sshd')
  end

  it 'creates the directory /etc/ssh' do
    expect(chef_run).to create_directory('/etc/ssh')
      .with(mode: '0755')
      .with(owner: 'root')
      .with(group: 'root')
  end

  it 'disables weak hmacs' do
    expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
      .with_content(/MACs [^#]*\bhmac-sha1\b/)
  end

  it 'disables weak kexs' do
    expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
      .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
  end

  it 'disables cbc ciphers' do
    expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
      .with_content(/Ciphers [^#]*-cbc\b/)
  end

  it 'enables ctr ciphers' do
    expect(chef_run).to render_file('/etc/ssh/sshd_config')
      .with_content(/Ciphers [^#]*\baes128-ctr\b/)
      .with_content(/Ciphers [^#]*\baes192-ctr\b/)
      .with_content(/Ciphers [^#]*\baes256-ctr\b/)
  end

  context 'with weak hmacs enabled' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['ssh']['weak_hmac'] = true
      end.converge(described_recipe)
    end

    it 'allows weak hmacs' do
      expect(chef_run).to render_file('/etc/ssh/sshd_config')
        .with_content(/MACs [^#]*\bhmac-sha1\b/)
    end

    it 'still does not allow weak kexs' do
      expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
        .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
    end

    it 'still doss not allow cbc ciphers' do
      expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
        .with_content(/Ciphers [^#]*-cbc\b/)
    end
  end

  context 'with weak kexs enabled' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['ssh']['weak_kex'] = true
      end.converge(described_recipe)
    end

    it 'allows weak kexs' do
      expect(chef_run).to render_file('/etc/ssh/sshd_config')
        .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
    end

    it 'still does not allow weak macs' do
      expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
        .with_content(/MACs [^#]*\bhmac-sha1\b/)
    end

    it 'still does not allow cbc ciphers' do
      expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
        .with_content(/Ciphers [^#]*-cbc\b/)
    end
  end

  context 'with cbc required' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['ssh']['cbc_required'] = true
      end.converge(described_recipe)
    end

    it 'allows cbc ciphers' do
      expect(chef_run).to render_file('/etc/ssh/sshd_config')
        .with_content(/Ciphers [^#]*\baes256-cbc\b/)
        .with_content(/Ciphers [^#]*\baes192-cbc\b/)
        .with_content(/Ciphers [^#]*\baes128-cbc\b/)
    end

    it 'still does not allow weak macs' do
      expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
        .with_content(/MACs [^#]*\bhmac-sha1\b/)
    end

    it 'still does not allow weak kexs' do
      expect(chef_run).not_to render_file('/etc/ssh/sshd_config')
        .with_content(/KexAlgorithms [^#]*\bdiffie-hellman-group1-sha1\b/)
    end

    it 'still enables ctr ciphers' do
      expect(chef_run).to render_file('/etc/ssh/sshd_config')
        .with_content(/Ciphers [^#]*\baes128-ctr\b/)
        .with_content(/Ciphers [^#]*\baes192-ctr\b/)
        .with_content(/Ciphers [^#]*\baes256-ctr\b/)
    end
  end

  it 'restarts the ssh server on config changes' do
    resource = chef_run.template('/etc/ssh/sshd_config')
    expect(resource).to notify('service[sshd]').to(:restart).delayed
  end

  it 'creates .ssh directory for user root' do
    expect(chef_run).to create_directory('/root/.ssh')
      .with(mode: '0500')
      .with(owner: 'root')
      .with(group: 'root')
  end

  context 'without attribute allow_root_with_key' do
    it 'does not unlock root account' do
      expect(chef_run).to_not run_execute('unlock root account if it is locked')
    end
  end

  context 'with attribute allow_root_with_key' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['ssh']['allow_root_with_key'] = true
      end.converge(described_recipe)
    end

    it 'unlocks root account' do
      expect(chef_run).to run_execute('unlock root account if it is locked')
        .with(command: "sed 's/^root:\!/root:*/' /etc/shadow -i")
    end
  end

  context 'with users data bag' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new do |_node, server|
        server.create_data_bag(
        'users',
        'user1' => { id: 'user1', ssh_rootkey: 'key-user1' },
        'user2' => { id: 'user2', ssh_rootkey: 'key-user2' },
        'user3' => { id: 'user3', ssh_rootkeys: %w(key1-user3 key2-user3) },
        'user4' => { id: 'user4', ssh_rootkeys: %w(key1-user4) }
      )
      end.converge(described_recipe)
    end

    it 'creates authorized_keys for root' do
      expect(chef_run).to create_template('/root/.ssh/authorized_keys')
        .with(mode: '0400')
        .with(owner: 'root')
        .with(group: 'root')
    end

    it 'authorizes keys from the user data bag for root access' do
      expect(chef_run).to render_file('/root/.ssh/authorized_keys')
        .with_content(/^key-user1$/)
        .with_content(/^key-user2$/)
        .with_content(/^key1-user3$/)
        .with_content(/^key2-user3$/)
        .with_content(/^key1-user4$/)
    end

    it 'warns about deprecation of data bag use' do
      expect(chef_run).to write_log('deprecated-databag')
        .with(message: /deprecated/)
        .with(level: :warn)
    end
  end

  context 'without users data bag' do
    cached(:chef_run) do
      ChefSpec::ServerRunner.new.converge(described_recipe)
    end

    it 'does not raise an error' do
      expect { chef_run }.not_to raise_error
    end

    it 'does not touch authorized_keys by root' do
      expect(chef_run).to_not create_template('/root/.ssh/authorized_keys')
    end

    it 'does not warn about deprecation of data bag use' do
      expect(chef_run).not_to write_log('deprecated-databag')
    end
  end

  context 'chef-solo' do
    cached(:chef_run) do
      ChefSpec::SoloRunner.new.converge(described_recipe)
    end

    it 'does not raise an error' do
      expect { chef_run }.not_to raise_error
    end
  end
end


recipes/server.rb
# encoding: utf-8
#
# Cookbook Name:: ssh-hardening
# Recipe:: server.rb
#
# Copyright 2012, Dominik Richter
# Copyright 2014, Deutsche Telekom AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# installs package name
package 'openssh-server' do
  package_name node['sshserver']['package']
end

# defines the sshd service
service 'sshd' do
  # use upstart for ubuntu, otherwise chef uses init
  # @see http://docs.opscode.com/resource_service.html#providers
  case node['platform']
  when 'ubuntu'
    if node['platform_version'].to_f >= 12.04
      provider Chef::Provider::Service::Upstart
    end
  end
  service_name node['sshserver']['service_name']
  supports value_for_platform(
    'centos' => { 'default' => [:restart, :reload, :status] },
    'redhat' => { 'default' => [:restart, :reload, :status] },
    'fedora' => { 'default' => [:restart, :reload, :status] },
    'scientific' => { 'default' => [:restart, :reload, :status] },
    'arch' => { 'default' => [:restart] },
    'debian' => { 'default' => [:restart, :reload, :status] },
    'ubuntu' => {
      '8.04' => [:restart, :reload],
      'default' => [:restart, :reload, :status]
    },
    'default' => { 'default' => [:restart, :reload] }
  )
  action [:enable, :start]
end

directory '/etc/ssh' do
  mode '0755'
  owner 'root'
  group 'root'
  action :create
end

template '/etc/ssh/sshd_config' do
  source 'opensshd.conf.erb'
  mode '0600'
  owner 'root'
  group 'root'
  variables(
    mac: SshMac.get_macs(node, node['ssh']['weak_hmac']),
    kex: SshKex.get_kexs(node, node['ssh']['weak_kex']),
    cipher: SshCipher.get_ciphers(node, node['ssh']['cbc_required'])
  )
  notifies :restart, 'service[sshd]'
end

def chef_solo_search_installed?
  klass = ::Search.const_get('Helper')
  return klass.is_a?(Class)
rescue NameError
  return false
end

# authorized_key management will be deprecated in the next major release:
def get_key_from(field)
  return [] if Chef::Config[:solo] && !chef_solo_search_installed?
  return [] unless Chef::DataBag.list.key?('users')
  search('users', "#{field}:*").map do |v| # ~FC003 ignore footcritic violation
    Chef::Log.info "ssh_server: installing ssh-keys for root access of user #{v['id']}"
    v[field]
  end.flatten
end

keys = get_key_from('ssh_rootkey') + get_key_from('ssh_rootkeys')

directory '/root/.ssh' do
  mode '0500'
  owner 'root'
  group 'root'
  action :create
end

unless keys.empty?
  log 'deprecated-databag' do
    message 'Use of deprecated key ssh_rootkey(s) found in users data bag. ' \
      'Managing authorized_keys from users data bag will be removed ' \
      'from the ssh-hardening cookbook in the next major release. ' \
      'Please transition to alternative approaches.'
    level :warn
  end

  template '/root/.ssh/authorized_keys' do
    source 'authorized_keys.erb'
    mode '0400'
    owner 'root'
    group 'root'
    variables(
      keys: keys
    )
  end
end

execute 'unlock root account if it is locked' do
  command "sed 's/^root:\!/root:*/' /etc/shadow -i"
  only_if { node['ssh']['allow_root_with_key'] }
end
